src/java/org/apache/cassandra/db/SinglePartitionReadCommand.java

```diff
@@ -1042,16 +1042,26 @@ private boolean canRemoveRow(Row row, Columns requestedColumns, long sstableTime
         // We can remove a row if it has data that is more recent that the next sstable to consider for the data that the query
         // cares about. And the data we care about is 1) the row timestamp (since every query cares if the row exists or not)
         // and 2) the requested columns.
-        if (row.primaryKeyLivenessInfo().isEmpty() || row.primaryKeyLivenessInfo().timestamp() <= sstableTimestamp)
+        
+        // Note that compact tables will always have an empty primary key liveness info. 
+        if (!row.primaryKeyLivenessInfo().isEmpty() && row.primaryKeyLivenessInfo().timestamp() <= sstableTimestamp) 
             return false;
-
+        
+        boolean hasLiveCell = false;
+        
         for (ColumnDefinition column : requestedColumns)
         {
             Cell cell = row.getCell(column);
-            if (cell == null || cell.timestamp() <= sstableTimestamp)
+            
+            if (cell == null || cell.timestamp() <= sstableTimestamp) 
                 return false;
+            
+            if (!cell.isTombstone()) 
+                hasLiveCell = true;
         }
-        return true;
+
+        // If we've gotten here w/ a compact table or at least one non-tombstone cell, it's safe to remove the row.
+        return hasLiveCell || !metadata().isCQLTable();
```

Comment:
After dropping compact storage, we may have to look at more SSTable than logically necessary, but for tables that never used compact storage, we have to look at older SSTables for primary key liveness info if all we have here are tombstones. Not looking at the older ones would mean we propagate empty primary key liveness info into the response creation, which would cause us to return zero rows instead of rows w/ null cell values.

Reply:
I'm struggling to read which of the unit tests capture that^. Do any of them?

Reply:
@michaelsembwever The easiest way to see this is to change this line to `return true;` and then run `operations.UpgradeTest` and `operations.DeleteTest`. You'll see failures around the issue above. Switching to `return hasLiveCell;` will eliminate that particular problem. Then run `SSTablesIteratedTest`, and you'll see one failure around `testCompactTableCellDeletion` hitting too many SSTables before DROP CS blows away the `isDense` flag. Adding `|| !metadata().isCQLTable()` resolves that. (i.e. When the table is compact, returning no rows rather than is perfectly acceptable. See the [comment below](https://github.com/apache/cassandra/pull/823/files?file-filters%5B%5D=.java#r525275385).)